### PR TITLE
refactor: validate only new entries on convenience-mode writes (#1086)

### LINKED ISSUE #1086: [FEATURE] Decouple add_source local shape check from pre-existing energy_sources entries
### What would you like to see?

In `_add_source` (PR #1073), the convenience-mode write path goes through `_mutate_atomic`, which builds a `partial_config = {"energy_sources": existing + [source]}` and submits it to `_set_prefs`. `_set_prefs` then runs `_shape_check` on the full combined payload before writing ([`tools_energy.py:574`](https://github.com/homeassistant-ai/ha-mcp/blob/802ac6691c1c0ce0b7680307cb8dca8ebfc19a3c/src/ha_mcp/tools/tools_energy.py#L574)).

This means a **pre-existing entry** must pass the local shape check on every `add_source` call — even though that entry is HA-validated state we should be able to trust.

The asymmetry is currently invisible because the local `_shape_check` for `energy_sources` is a strict subset of HA's voluptuous schema (only `type ∈ {grid, solar, battery, gas}` plus `stat_energy_from` for non-grid types — both also required by HA). So any entry HA accepted will pass our check.

But if `_shape_check` is ever tightened to be stricter than HA in some respect — mandatory `name`, type-specific cost-config rules, unit normalisation, etc. — `add_source` would fail when appending an *unrelated* new source, because a *pre-existing* entry would now fail the stricter local check. The user would see their valid HA config rejected with no insight into why.

The structural fix is to scope `add_source`'s shape check to *just the new source*, not the union with the existing list.

### Why do you need this?

Forward-looking robustness. No reachable bug today (verified live on HA 2026.4 against current `_shape_check`). But the asymmetric coupling means any future tightening of `_shape_check` quietly breaks the ability to append until the user manually fixes their existing config — which is poor UX.

Filed as a forward-looking note from the [PR #1073 self-review](https://github.com/homeassistant-ai/ha-mcp/pull/1073#issuecomment-4360817307) (O2). Not a blocker for that PR.

### Additional context

`_add_source` already does a per-entry shape check on the new source alone before delegating to `_mutate_atomic` ([`tools_energy.py:843`](https://github.com/homeassistant-ai/ha-mcp/blob/802ac6691c1c0ce0b7680307cb8dca8ebfc19a3c/src/ha_mcp/tools/tools_energy.py#L843)) — so the new source is already validated independently. The redundant validation happens later in `_set_prefs`, on the union.

**Maintainer choice required — at least three candidate shapes:**

1. **Skip the second shape check in `_set_prefs` for convenience-mode payloads.** Plumb a flag (or a new `_set_prefs_unchecked` variant) so `_mutate_atomic` callers — which have already validated their own input — can opt out of the union shape check. Smallest change. Drawback: now there are two write paths into `_set_prefs` with different validation behaviour.

2. **Validate only new entries via a `_shape_check` mode flag.** Add a `validate_only_keys: set[str] | None` (or per-key indices) parameter so `_set_prefs` validates only the new entries inside each top-level list, not the union. More general — covers `add_device`/`remove_device` symmetrically (`remove_device` doesn't need *any* per-entry validation, since it's just reducing the list). Drawback: bigger refactor, more API surface on a private helper.

3. **Drop the `_set_prefs` shape check entirely, trust `_dry_run` / per-entry checks.** `_set_prefs` would rely on the caller (or `_dry_run`) to have validated, plus HA's own server-side validation as the final gate. Drawback: loses the safety net for direct `mode='set'` callers who skipped `dry_run`.

**Code references** (pinned to PR #1073 HEAD `802ac669`):

- `_shape_check()` def — [L124](https://github.com/homeassistant-ai/ha-mcp/blob/802ac6691c1c0ce0b7680307cb8dca8ebfc19a3c/src/ha_mcp/tools/tools_energy.py#L124)
- `_set_prefs` shape-check call — [L574](https://github.com/homeassistant-ai/ha-mcp/blob/802ac6691c1c0ce0b7680307cb8dca8ebfc19a3c/src/ha_mcp/tools/tools_energy.py#L574)
- `_add_source` per-entry shape check — [L843](https://github.com/homeassistant-ai/ha-mcp/blob/802ac6691c1c0ce0b7680307cb8dca8ebfc19a3c/src/ha_mcp/tools/tools_energy.py#L843)
- `_mutate_atomic` partial-config build — [L970-972](https://github.com/homeassistant-ai/ha-mcp/blob/802ac6691c1c0ce0b7680307cb8dca8ebfc19a3c/src/ha_mcp/tools/tools_energy.py#L970-L972)

Reproducible only after PR #1073 lands — `_add_source` and `_mutate_atomic` are introduced by it.

### PR DESCRIPTION
**Pull request by Patch76**
## What does this PR do?

Closes #1086. Convenience-mode writes (`_add_device`, `_add_source`, `_remove_device`) build a `partial_config` equal to the FULL existing list ± one entry, then route it through `_set_prefs` which ran `_shape_check` on that union — re-validating pre-existing (HA-validated) siblings on every add/remove. If `_shape_check` is ever tightened past HA's voluptuous schema, an unrelated add would fail because a SIBLING entry would now fail the local check.

Scope `_shape_check` on the convenience-mode write path to just the appended tail:

- `_shape_check` accepts `validate_only: dict[str, set[int]] | None`. `None` preserves the original full-validation contract. An empty dict skips the entire per-entry pass. Structural "must be a dict" / "must be a list" checks still fire for listed keys.
- `_set_prefs` accepts and forwards `validate_only` as a kwarg-only param.
- `_mutate_atomic` computes `appended_indices = range(len(existing_list), len(new_list))` and passes `{target_key: appended_indices}` on both the real-run write and the dry-run backstop. For `add_*` this is the new entry only; for `remove_*` this is `set()` so nothing is re-validated.
- Direct `mode='set'` callers are **unchanged** — `validate_only` defaults to `None` and the full check still runs.

Assumes append-only/remove-only mutator semantics; revisit if an in-place mutator is ever added.

**No reachable bug on master today** — the local `_shape_check` is currently a strict subset of HA's voluptuous schema, so any HA-accepted entry passes the local check. This is forward-looking robustness; see issue body for the full motivation.

Touches `_set_prefs` alongside the in-flight #1098 (per-key `config_hash`). Both add kwarg-only params to the same signature; otherwise distinct validation logic. Whichever lands second rebases trivially.

## Type of change
- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 📚 Documentation
- [x] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [ ] All automated tests pass (`uv run pytest`) — local Alpine env lacks Python 3.13; will check after CI green
- [x] Code follows style guidelines (`uv run ruff check`) — repo-pinned ruff 0.15.12 clean on changed files

Tests added:

- `TestShapeCheckValidateOnly` (7 unit tests): `validate_only=None` parity, empty-dict full skip, unlisted-key skip, unlisted-index skip within a key, listed-bad-index still raises, structural "must be a list" check fires for listed keys but skipped for unlisted keys.
- `TestConvenienceModesPreExistingInvalid` (5 regression tests): `add_device` / `add_source` / `remove_device` succeed despite a deliberately-invalid pre-existing sibling; dry-run mirrors real-run; direct `mode='set'` still rejects an invalid config (boundary).

## Checklist
- [x] I have updated documentation if needed — `_shape_check`'s and `_set_prefs`'s docstrings document `validate_only`; `_mutate_atomic`'s real-run and dry-run paths carry inline notes pinning the heuristic to #1086.

### CODE REVIEW

File: src/ha_mcp/tools/tools_energy.py [hunk 1]
```diff
@@ -121,13 +121,28 @@ def _flatten_validation_errors(raw: Any) -> list[dict[str, str]]:
...
+    every entry under every present top-level key — the original contract.
+    A dict scopes the check to the listed keys and, within each, only the
+    listed indices: top-level keys absent from the dict are skipped entirely,
+    indices outside each key's set are skipped per-entry. The "must be a
+    list" structural check still fires for any present-and-listed key with a
+    non-list value, so ``validate_only`` cannot be used to bypass structural
+    sanity. An empty dict skips the entire per-entry pass (used by
+    convenience-mode write paths whose mutator did not append any new
+    entries — e.g. remove operations). See issue #1086 for the asymmetric
+    over-validation problem this addresses.
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_energy.py]: Fixed in `c71e372`. The docstring now distinguishes `{key: set()}` (skips per-entry pass for that key, preserves the structural "must be a list" check — what convenience-mode passes for removals) from `{}` (skips all keys entirely). Code semantics were correct; only the docstring text needed the correction.

**Comment by Patch76** [comment]
## Implementation Summary

**Choices Made:**
- **`validate_only: dict[str, set[int]] | None` parameter on `_shape_check`** — `None` keeps the legacy full-check contract; a dict scopes validation per-top-level-key and per-index. Lets `_set_prefs` skip pre-existing siblings when called from `_mutate_atomic`, where the caller knows which entries are new. Picked Shape 2 from the issue body over Shape 1 (skip-flag): per-key-indices is more general — covers `_add_device`, `_add_source`, AND `_remove_device` symmetrically (Shape 1 would have masked validation of new entries on the convenience path), and Shape 2 still validates the genuinely new entry inside `_set_prefs` rather than trusting the caller-side pre-check entirely.
- **Per-key allowed-indices set, not a global indices set** — keys absent from `validate_only` are skipped entirely; keys present with empty sets skip the per-entry pass for that key while preserving the structural "must be a list" check (Gemini G1, adopted; clarified in `58e1f6f`).
- **`_mutate_atomic` computes `appended_indices = set(range(len(existing_list), len(new_list)))`** — handles all three convenience modes uniformly: `add_*` produces a singleton `{N}`, `remove_*` produces `set()`. Heuristic assumes append-only/remove-only mutator semantics; revisit if an in-place mutator is ever added.
- **Both real-run write and dry-run backstop scoped equivalently** — `_mutate_atomic`'s dry-run path also calls `_shape_check` with the same `validate_only={target_key: appended_indices}`. Keeps the two paths shape-equivalent (matches the inline comment at `tools_energy.py:1243-1248`).
- **Direct `mode='set'` callers unchanged** — `validate_only` defaults to `None`, so explicit-power-user writes still get the full check.
- **Helper docstrings start with action verbs** (Gemini G2, adopted in `fa831bb`) — applied across `_shape_check` ("Perform a cheap..."), `_is_no_prefs_error` ("Return True if..."), and `_mutate_atomic` ("Run the read-modify-write loop...") for consistency with the file's empirical 11/15 action-verb-first pattern (and the project-wide pattern in `tools_addons.py` / `tools_integrations.py`). Note: the `.gemini/styleguide.md` rule itself is scoped to public `@tool` docstrings, so adoption here is empirical-pattern-consistency rather than literal styleguide-compliance.
- **Self-Review Boy-Scout sweep** (adopted in `35a7209`) — three docstring tightenings on the lines this PR already touches: `_shape_check` first line dropped the redundant "cheap"+"shape check" phrasing ("Perform a cheap..." → "Validate config shape locally..."); `_mutate_atomic` lead expanded from "Run the read-modify-write loop for convenience modes" to "Run convenience-mode read-modify-write with dry-run backstop and hash-conflict retry" (matches the function's actual responsibilities at L1313-1336); the heuristic warning at the `appended_indices` computation (real-run + dry-run) made the in-place-mutator failure mode explicit ("len(new) == len(existing) but content changed" → empty `appended_indices`, skips per-entry pass entirely).

**Problems Encountered:**
- **Gemini G1: `_shape_check` docstring conflated `{}` and `{key: set()}`** (`tools_energy.py:164`, medium) — the prior wording said `{}` was what convenience-mode passed for removals, but the actual code at `_mutate_atomic` passes `{target_key: set()}`. Code semantics were already correct; only the docstring needed the precision. Addressed in `58e1f6f` by spelling out both shapes.
- **Rebase conflict on `_set_prefs` after #1098 merged** — master gained `_set_prefs.config_hash: str | dict[_PrefsKey, str]` and a per-key-hash validation block from #1098 while #1100 was Draft. The source-file conflict (`_set_prefs` docstring) merged trivially. The test-file conflict was awkwardly interleaved — both branches added new test classes at adjacent insertion points sharing common scaffolding lines (`{"success": True, "result": None}` etc.), so git's auto-merger couldn't isolate the additions. Resolved by reconstructing the test file from master's version with my two new classes (`TestShapeCheckValidateOnly` after `TestShapeCheck`, `TestConvenienceModesPreExistingInvalid` after `TestConvenienceModesPassStrHash`) inserted at structurally-clean class boundaries. Force-pushed; both prior threads (`PRRT_kwDOPvq5vc5_Mxrh`, `PRRT_kwDOPvq5vc5_M_mu`) survived as `isResolved: true`. PR is now `mergeable: MERGEABLE`.
- **Leftover merge-conflict markers in `_set_prefs` docstring** — the rebase resolution dropped `<<<<<<< HEAD` / `=======` / `>>>>>>> 0825a15` markers but the `>>>>>>>`-line and the diff between the two halves slipped past `git rebase --continue` because the markers sat inside a triple-quoted docstring. Pure-string content; ruff, mypy, AST parsing, and pytest 1479/1 all pass with markers in place. Caught pre-merge and addressed in `426854d` (no behaviour change; merged the two halves into one cohesive paragraph).

**Suggested Improvements:**
- None — all review items addressed inline; no follow-up issues filed.

Rebased onto master post-#1098 (per-key `config_hash`, merged at `8533c8c`). Both PRs added kwarg-only params to `_set_prefs`'s signature plus orthogonal validation logic; the merge of the two plus pre-merge fix-ups is what's now on this branch (commits `278891a` / `58e1f6f` / `fa831bb` / `426854d` / `35a7209`).

Local pre-push validation: full ha-mcp unit suite 1479 passed, 1 skipped (`test_numpy_version_constraint` — numpy is a transitive optional dependency, not resolved in this venv), 34.71s on Python 3.13.13. Mypy + Ruff clean.

File: src/ha_mcp/tools/tools_energy.py [hunk 2]
```diff
@@ -121,13 +121,30 @@ def _flatten_validation_errors(raw: Any) -> list[dict[str, str]]:
     return errors
 
 
-def _shape_check(config: dict[str, Any]) -> list[dict[str, str]]:
+def _shape_check(
+    config: dict[str, Any],
+    validate_only: dict[str, set[int]] | None = None,
+) -> list[dict[str, str]]:
     """Cheap local shape check before sending to the server.
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_energy.py]: Adopted in `ef75ab5`. Applied across `_shape_check`, `_is_no_prefs_error`, and `_mutate_atomic` for consistency with the file's empirical 11/15 action-verb-first pattern (and the project-wide pattern in `tools_addons.py`/`tools_integrations.py`). Note: the `.gemini/styleguide.md` rule itself is scoped to public `@tool` docstrings (verb whitelist `Get`/`List`/`Search`/`Create`/`Update`/`Delete`/`Remove`/`Execute`/`Call`/`Manage`), so adoption here is empirical-pattern-consistency rather than literal styleguide-compliance — but the empirical pattern in tool modules clearly extends to internal helpers.

**Comment by Patch76** [comment]
Self-Review on this PR (pre-Ready audit) flagged three docstring tightenings worth folding in (Boy-Scout, addressed in `35a7209`) and caught three leftover `<<<<<<< HEAD` / `=======` / `>>>>>>>` markers in `_set_prefs`'s docstring that had survived the post-#1098 rebase — they sat inside the triple-quoted string, so ruff, mypy, AST parsing, and pytest 1479/1 all passed with markers in place. Addressed in `426854d` (no behaviour change; merged the two halves into one paragraph).

The IS comment above has been updated in-place: new Choices-Made bullet for the Boy-Scout sweep, new Problems-Encountered bullet for the marker-fix, plus the commit list at the bottom now lists all five.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
Two real changes wanted, plus four cleanups.

**Requested**

1. **Convert the in-place-mutator assumption into a runtime guard** — `tools_energy.py:1244-1248`, `:1293-1298`. The "revisit before adding such a mutator" comment is the only forward-looking exposure: a future `_replace_*` / `_update_*` mutator with `len(new) == len(existing)` would yield `appended_indices = set()` and silently skip every per-entry check on a modified entry — exactly the validation asymmetry #1086 was meant to address, just inverted. Either an assertion that fails on same-length non-equal mutations, or a registered mutator-kind that `_mutate_atomic` matches on, so the wrong shape fails loudly at test time rather than in production.

2. **Replace the unreachable-loop `AssertionError` with a structured error** — `tools_energy.py:1352-1355`. If the optimistic-concurrency loop ever does exhaust without taking either branch, the bare `AssertionError` falls through `except Exception` at 1359 and surfaces as a generic `INTERNAL_ERROR` with no actionable suggestions. `raise_tool_error(create_error_response(ErrorCode.INTERNAL_ERROR, ..., suggestions=[...]))` matches the pattern used elsewhere in this file.

**While you're in there**

3. Extract `_appended_tail_indices(existing, new)` so the 8-line "append-only / shrink-only" rationale lives in one place. Currently duplicated verbatim at `:1239-1248` and `:1289-1298` — DRY plus drift risk if one site updates.

4. `test_add_device_still_rejects_a_genuinely_invalid_new_entry` (`test_tools_energy.py:1314`) actually exercises `mode='set'`, not `add_device`. Rename to e.g. `test_direct_set_mode_still_validates_full_config_when_validate_only_is_none` so the boundary it pins is in the name.

5. Add a direct `_shape_check` test for `validate_only={key: set()}` on a non-empty list with bad entries — pins the docstring's distinction between `{}` (skip everything) and `{key: set()}` (key present, structural list-check still fires, per-entry skipped). Currently only `{}` is unit-tested at `_shape_check` level; `{key: set()}` is exercised end-to-end via `remove_device` but not directly.

6. Parametrize `test_*_dry_run_succeeds_with_invalid_pre_existing` across `add_device` / `add_source` / `remove_device`. The dry-run code path at `_mutate_atomic:1233-1268` is shared so one test gives reasonable confidence, but symmetry with the real-run regression block hardens against future divergence.

Tests, docstrings, and the validate_only contract itself are all solid.

**Comment by Patch76** [comment]
✅ Addressed all 6 review items in `18d5578`.

**Requested:**

- **R1 — Runtime guard for in-place-mutator assumption** — extracted `_appended_tail_indices(existing, new)` at `tools_energy.py:244` and added an `INTERNAL_ERROR` raise when `len(new) == len(existing) and new != existing`. A future `_replace_*` / `_update_*` mutator now fails loudly at the helper rather than silently bypassing per-entry validation.
- **R2 — Structured error replaces `AssertionError`** at the post-loop unreachable point (`tools_energy.py:1379-1390`). Now raises `raise_tool_error(create_error_response(ErrorCode.INTERNAL_ERROR, ..., context={"mode": mode, "target_key": target_key}, suggestions=[...]))` so the surface matches the rest of this file rather than falling through `except Exception` to a context-less generic.

**While you're in there:**

- **W3 — DRY** — duplicated 8-line "append-only / shrink-only" rationale at the dry-run + real-run sites collapsed to one-liner calls of `_appended_tail_indices`. Rationale now lives in the helper docstring; both call sites carry a 4-line ref-comment instead of 10 lines of duplicated prose.
- **W4 — Test renamed** — `test_add_device_still_rejects_a_genuinely_invalid_new_entry` → `test_direct_set_mode_still_validates_full_config_when_validate_only_is_none`. Body unchanged; only the name + docstring now describe the actual boundary it pins (the direct `mode='set'` path keeps `validate_only=None` default and full-validates).
- **W5 — Direct `_shape_check` test for `validate_only={key: set()}`** — added `test_listed_key_with_empty_index_set_skips_per_entry_but_keeps_structural` in `TestShapeCheckValidateOnly`. Pins the docstring's `{}` (skip everything) vs `{key: set()}` (key listed, list-shape still checked, per-entry skipped) distinction at the unit level rather than only via the end-to-end `remove_device` path.
- **W6 — Parametrized dry-run test** — converted `test_add_device_dry_run_succeeds_with_invalid_pre_existing` to `@pytest.mark.parametrize` covering `add_device` / `add_source` / `remove_device`. Symmetry with the real-run regression block above so future divergence between `_mutate_atomic`'s two branches breaks loudly.

Local: 101/101 pytest, ruff + mypy clean.

**Review by kingpanther13 — APPROVED** [review]
All six items addressed in `18d5578`. `_appended_tail_indices` reads cleanly, the in-place-mutator guard fails loudly with actionable context, and the structured `INTERNAL_ERROR` at the post-loop exit is a strict upgrade over the bare `AssertionError`. Parametrized dry-run + the direct `{key: set()}` unit test close the symmetry gap nicely.

CI green, no conflicts. LGTM.
